fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,639 @@
+                - 2 * total_active_balance
+            ) // (3 * remaining_ffg_weight)
+        )
+        if min_ffg_support_slash_th * 100 <= total_active_balance * ffg_confirmation_score:
+            return ffg_confirmation_score
+    else:
+        if 2 * total_active_balance <= (ffg_support_for_checkpoint - min_ffg_support_slash_th) * 3:
+            return 100 // 3
+
+    # Case 2: min_ffg_support_slash_th > total_active_balance * ffg_confirmation_score / 100
+    #
+    # In this case the problem reduces to returning the max possible ffg_confirmation_score such that:
+    #
+    # 2 / 3 * total_active_balance <= \
+    #     ffg_support_for_checkpoint - \
+    #     total_active_balance * ffg_confirmation_score / 100 + \
+    #     (1 - ffg_confirmation_score / 100) * remaining_ffg_weight
+
+    ffg_confirmation_score = (
+        100 * (3 * (ffg_support_for_checkpoint + remaining_ffg_weight) - 2 * total_active_balance) //
+        (3 * (remaining_ffg_weight + total_active_balance))
+    )
+    assert ffg_support_for_checkpoint * 100 > total_active_balance * ffg_confirmation_score
+    return max(ffg_confirmation_score, UNCONFIRMED_SCORE)
+```
+
+#### `get_ffg_confirmation_score`
+
+```python
+def get_ffg_confirmation_score(store: Store, block_root: Root) -> int:
+
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    total_active_balance = int(get_total_active_balance_for_block_root(store, block_root))
+
+    ffg_support_for_checkpoint = int(get_ffg_support(store, block_root))
+
+    min_ffg_support_slash_th = min(ffg_support_for_checkpoint, CONFIRMATION_SLASHING_THRESHOLD)
+
+    if block_epoch == current_epoch:
+        remaining_ffg_weight = int(get_remaining_weight_in_current_epoch(store, block_root))
+    else:
+        remaining_ffg_weight = 0  
+
+    return get_ffg_confirmation_score_helper(
+        total_active_balance,
+        ffg_support_for_checkpoint,
+        min_ffg_support_slash_th,
+        remaining_ffg_weight
+    )          
+```
+
+### `get_confirmation_score`
+
+```python
+def get_confirmation_score(store: Store, block_root: Root) -> int:
+    """
+    Return `UNCONFIRMED_SCORE` in the case that `block_root` cannot be confirmed even by assuming no adversary weight,
+    otherwise it returns the maximum percentage of adversary weight that is admissible in order to
+    consider `block_root` confirmed.
+    """
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
```

Comment:
Assuming `current_epoch - 2, current_epoch - 1` are finalized and justified respectively and the synchrony assumption is held, consider a block from `current_epoch - 2` that is older than the finalized block. This block is very likely confirmed, while the API using this function call will have to do some additional computation to prove that. If a block is in the finalized chain the lookup can be simple while for those that aren't finalized the thing is more complicated.

This restriction can cause a weird UX when a previously confirmed status for a block turns into a void status. Clients will have to do something about to prevent this hiccup.

Reply:
This is definitely an issue that needs to be fixed.
What we can do for a block `B` from an epoch `e < current_epoch - 1` is to check whether the first descendant of `B` from an epoch `>= current_epoch - 1` is confirmed.

Reply:
Added the logic above.

Reply:
Added similar logic for the confirmation score as well.